tests/std/tests/P1209R0_erase_if_erase/test.cpp

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cassert>
+#include <deque>
+#include <forward_list>
+#include <list>
+#include <map>
+#include <set>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+template <class>
+constexpr bool is_unsized_container = false;
+template <class T, class A>
+constexpr bool is_unsized_container<std::forward_list<T, A>> = true;
+
+// Note that the standard actually requires these to be copyable. As an extension, we want to ensure we don't copy them,
+// because copying some functors (e.g. std::function) is comparatively expensive, and even for relatively cheap to copy
+// function objects we care (somewhat) about debug mode perf.
+struct no_copy {
+    no_copy()                          = default;
+    no_copy(const no_copy&)            = delete;
+    no_copy(no_copy&&)                 = default;
+    no_copy& operator=(const no_copy&) = delete;
+    no_copy& operator=(no_copy&&)      = delete;
+};
+
+struct is_vowel : no_copy {
+    constexpr bool operator()(const char c) const {
+        return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u';
+    }
+};
+
+struct is_odd : no_copy {
+    constexpr bool operator()(const int i) const {
+        return i % 2 != 0;
+    }
+};
+
+struct is_first_odd : no_copy {
+    bool operator()(const std::pair<const int, int>& p) const {
+        return p.first % 2 != 0;
+    }
+};
+
+constexpr bool test_string() {
+    std::string str1{"cute fluffy kittens"};
+    const auto str1_removed = std::erase_if(str1, is_vowel{});
+    assert(str1 == "ct flffy kttns");
+    assert(str1_removed == 5);
+
+    std::string str2{"asynchronous beat"};
+    const auto str2_removed = std::erase(str2, 'a');
+    assert(str2 == "synchronous bet");
+    assert(str2_removed == 2);
+
+    return true;
+}
+
+template <class SequenceContainer>
+constexpr bool test_sequence_container() {
+    SequenceContainer cont1{42, 1729};
+    const auto cont1_removed = std::erase(cont1, 42);
+    if constexpr (!is_unsized_container<SequenceContainer>) {
+        assert(cont1.size() == 1);
+    }
+    assert(cont1.front() == 1729);
+    assert(cont1_removed == 1);
+
+    SequenceContainer cont2{17, 42, 29};
+    const auto cont2_removed = std::erase_if(cont2, is_odd{});
+    if constexpr (!is_unsized_container<SequenceContainer>) {
+        assert(cont2.size() == 1);
+    }
+    assert(cont2.front() == 42);
+    assert(cont2_removed == 2);
+
+    return true;
+}
+
+// Also test LWG-4135 "The helper lambda of std::erase for list should specify return type as bool"
+
+template <bool B>
+struct pinned_condition {
+    pinned_condition()                                   = default;

```

Comment:
Suggestion: Like product code, we can make this `explicit` to defend against unintentional use.